test/cpu_test.cpp

```diff
@@ -669,6 +679,196 @@ TEST(cpu_test, convolution_large_padding)
     compare_backends(int_f, cpu_f, "INTERPRETER", "CPU", 1e-4, 1e-4);
 }
 
+TEST(cpu_test, memory_reuse_mxnet_densenet121)
+{
+    const std::string file_name("mxnet/mxnet_densenet121_inference_batch1_float32.json");
+    auto cpu_f = make_function(file_name);
+
+    test::Uniform<float> rng(-1.0f, 1.0f);
+    vector<vector<float>> args;
+
+    for (shared_ptr<op::Parameter> param : cpu_f->get_parameters())
+    {
+        vector<float> tensor_val(shape_size(param->get_shape()));
+        rng.initialize(tensor_val);
+        args.push_back(tensor_val);
+    }
+
+    // without memory reuse
+    auto cpu_results = execute(cpu_f, args, "CPU");
+
+    for (auto it = 0; it < 2; it++)
+    {
+        auto cpu_f_new = make_function(file_name);
+        auto cpu_results_new = execute(cpu_f_new, args, "CPU");
+        for (size_t i = 0; i < cpu_results.size(); i++)
+        {
+            EXPECT_TRUE(
+                test::all_close(cpu_results.at(i), cpu_results_new.at(i), 1.0e-4f, 1.0e-4f));
+        }
+    }
+
+    // with memory reuse
+    auto backend = runtime::Backend::create("CPU");
+    auto parms = cpu_f->get_parameters();
+    std::vector<std::shared_ptr<ngraph::runtime::Tensor>> arg_tensors(args.size());
+    for (size_t i = 0; i < args.size(); i++)
+    {
+        auto t = backend->create_tensor(parms.at(i)->get_element_type(), parms.at(i)->get_shape());
+        copy_data(t, args.at(i));
+        arg_tensors.at(i) = t;
+    }
+
+    auto results = cpu_f->get_results();
+    std::vector<std::shared_ptr<ngraph::runtime::Tensor>> result_tensors(results.size());
+
+    for (size_t i = 0; i < results.size(); i++)
+    {
+        result_tensors.at(i) =
+            backend->create_tensor(results.at(i)->get_element_type(), results.at(i)->get_shape());
+    }
+
+    ngraph::pass::PassConfig pass_config;
+    pass_config.set_reuse_memory(true);
+    auto cpu_backend = std::unique_ptr<runtime::cpu::CPU_Backend>(
+        static_cast<runtime::cpu::CPU_Backend*>(backend.release()));
+    for (auto it = 0; it < 2; it++)
+    {
+        auto cpu_f_new_reuse = make_function(file_name);
+
+        auto handle = cpu_backend->compile(cpu_f_new_reuse, pass_config);
+        cpu_backend->call_with_validate(handle, result_tensors, arg_tensors);
+
+        std::vector<std::vector<float>> cpu_results_new_reuse;
+        for (auto rt : result_tensors)
+        {
+            cpu_results_new_reuse.push_back(read_vector<float>(rt));
+        }
+
+        for (size_t i = 0; i < cpu_results.size(); i++)
+        {
+            EXPECT_TRUE(
+                test::all_close(cpu_results.at(i), cpu_results_new_reuse.at(i), 1.0e-4f, 1.0e-4f));
+        }
+    }
```

Comment:
this is quite an expensive test. We might want to disable by default and only run it selectively.